pybikes/pedalada.py

```diff
@@ -0,0 +1,72 @@
+import json
+
+from pybikes import BikeShareSystem, BikeShareStation, PyBikesScraper
+
+STATIONS_URL = '{endpoint}/php/get-stations-details-map.php'
+STATION_DETAILS_AND_RIDES_URL = '{endpoint}/php/get-station-details-and-rides.php?n={station_number}'
+STATION_IMAGE_URL = '{endpoint}/images/stations/station_{station_number}.jpg'
+
+
+def station_details_and_rides_url(endpoint, station_number):
+    return STATION_DETAILS_AND_RIDES_URL.format(endpoint=endpoint, station_number=station_number)
+
+def station_image_url(endpoint, station_number):
+    return STATION_IMAGE_URL.format(endpoint=endpoint, station_number=station_number)
+
+class Pedalada(BikeShareSystem):
+    meta = {
+        'company': [
+            'Share2Go - Mobilidade Partilhada, Sociedade Unipessoal Lda'
+        ]
+    }
+
+    def __init__(self, tag, meta, endpoint):
+        meta['company'] += Pedalada.meta['company']
+        super(Pedalada, self).__init__(tag, meta)
+        self.endpoint = endpoint
+
+    @property
+    def stations_url(self):
+        return STATIONS_URL.format(endpoint=self.endpoint)
+
+    def update(self, scraper=None):
+        scraper = scraper or PyBikesScraper()
+        stations_data = json.loads(scraper.request(self.stations_url))
+
+        stations = []
+        for station in stations_data:
+            station = PedaladaStation(station, self.endpoint)
+            stations.append(station)
+
+        self.stations = stations
+
+class PedaladaStation(BikeShareStation):
+    def __init__(self, data, endpoint):
+        super(PedaladaStation, self).__init__()
+        self.endpoint = endpoint
+
+        self.name = data['stationName']
+        self.latitude = data['latitude']
+        self.longitude = data['longitude']
+
+        self.extra = {
+            'uid': data['stationNumber'],
+            'slots': data['maximumNumberOfRides'],
+            'online': data['state'] == 1,
+            'photo': station_image_url(endpoint, data['stationNumber']),
+            'open': data['stationIsOpen'] == 1,
+            'in_maintenance': data['stationInMaintenance'] == 1
```

Comment:
Do you know the difference between `state`, `stationIsOpen` and `stationInMaintenance`?

In vitelinhas there's one station with the following:

```
{
  "name": "Avenida da Liberdade",
  "latitude": 41.45442,
  "longitude": -8.17702,
  "bikes": 3,
  "free": 8,
  "extra": {
    "uid": 2,
    "slots": 11,
    "online": false,
    "photo": "https://vitelinhas.cm-fafe.pt/images/stations/station_2.jpg",
    "open": true,
    "in_maintenance": false
  },
  "type": "station",
  "tag": "vitelinhas"
}
```

On the website that's reported as inactive. So my take on there is that open maybe means it's within the schedule opening hours?

Reply:
My take is that it's a bit confusing and we should try to find a better name for it, or decide to not include it if we skip schedules altogether

Reply:
> Do you know the difference between `state`, `stationIsOpen` and `stationInMaintenance`?
> 
> In vitelinhas there's one station with the following:
> 
> ```
> {
>   "name": "Avenida da Liberdade",
>   "latitude": 41.45442,
>   "longitude": -8.17702,
>   "bikes": 3,
>   "free": 8,
>   "extra": {
>     "uid": 2,
>     "slots": 11,
>     "online": false,
>     "photo": "https://vitelinhas.cm-fafe.pt/images/stations/station_2.jpg",
>     "open": true,
>     "in_maintenance": false
>   },
>   "type": "station",
>   "tag": "vitelinhas"
> }
> ```
> 
> On the website that's reported as inactive. So my take on there is that open maybe means it's within the schedule opening hours?

open means that the station would be open according to its schedule. But it can be unavailable due to connection problems but would be open per se. On the websites, the station being offline takes precedence over the schedules and if it is offline it's just shown as inactive. I believe a station in maintenance means that it is not available at the moment, due to being serviced. Maybe we could combine these all into one property.

Reply:
It's partly fault on pybikes side that we do not have a full specification for such cases. For some we have a `status` field that is a kind of all goes in. We could try to merge `open` and `in_maintenance` with the `status` field with one of the following `open`, `closed`, `maintenance`. Since we can consider that a station that is in maintenance is not open nor closed per se.

```
def format_status(self, data):
    if data['stationInMaintenance']:
        return 'maintenance'
    elif data['stationIsOpen']:
        return 'open'
    else:
        return 'closed'

[...]

'status': self.format_status(data),
```

Alternatively we can go with a oneliner but it's a bit of weird python

```
'status': 'maintenance' if data['stationInMaintenance'] else 'open' if data['stationIsOpen'] else 'closed'
```